sympy/matrices/matrices.py

```diff
@@ -1141,15 +1164,25 @@ def eigenvals(self, error_when_incomplete=True, **flags):
                     eigs[diagonal_entry] += 1
         else:
             flags.pop('simplify', None)  # pop unsupported flag
-            eigs = roots(mat.charpoly(x=Dummy('x')), **flags)
+            eigs = roots(mat.charpoly(x=Dummy('x'), simplify=simpfunc_1), **flags)
 
         # make sure the algebraic multiplicty sums to the
         # size of the matrix
         if error_when_incomplete and (sum(eigs.values()) if
             isinstance(eigs, dict) else len(eigs)) != self.cols:
             raise MatrixError("Could not compute eigenvalues for {}".format(self))
 
-        return eigs
+        # Since 'simplify' flag is unsupported in roots()
+        # simplify() function will be applied once at the end of the routine.
+        if not simplify:
+            return eigs
+        else:
```

Comment:
This could be

    if not isinstance(simplify, FunctionType):
        simplify = _simplify

Then we could use `simplify` instead of `simpfunc_2`. (`else:` and indentation is not necessary.)